### PR TITLE
LinalgRefactor - SGVector & SGMatrix - scale

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: src/shogun/mathematics/linalg/LinalgBackendBase.h [hunk 1]
```diff
@@ -104,6 +104,33 @@ class LinalgBackendBase
...
+	 * @see linalg::scale
+	 */
+	#define BACKEND_GENERIC_SCALE(Type) \
+	virtual SGVector<Type> scale(const SGVector<Type>& a, Type alpha) const \
+	{  \
+		SG_SNOTIMPLEMENTED; \
+	}
+
+	BACKEND_GENERIC_SCALE(bool);
+	BACKEND_GENERIC_SCALE(char);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: same comments as in other patch

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 2]
```diff
@@ -138,7 +150,9 @@ SGVector<T> from_gpu(const SGVector<T>& vector)
 		typedef typename std::aligned_storage<sizeof(T), alignof(T)>::type aligned_t;
 		T* data;
 		data = reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen));
+
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: pls no random whitespace changes

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 3]
```diff
@@ -111,6 +133,18 @@ class LinalgBackendEigen : public LinalgBackendBase
 		return (typename SGVector<T>::EigenVectorXtMap(a)).dot(typename SGVector<T>::EigenVectorXtMap(b));
 	}
 
+	/** Eigen3 vector scale method: B = alpha * A */
+	template <typename T>
+	SGVector<T> scale_impl(const SGVector<T>& a, T alpha) const
+	{
+		SGVector<T> b(a.vlen);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: We have to have a discussion on in-place vs allocating new memory

Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: We have to have a discussion on in-place vs allocating new memory

**Comment by karlnapf** [comment]
needs a rebase

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h (hunk 3 above)
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: I vote that for now, we *always* do these type of operations in place.
And if a developer doesnt want that, he can copy the vector first

**Comment by OXPHOS** [comment]
@vigsterkr ready to merge

**Pull request merged**